homeassistant/components/utility_meter/__init__.py

```diff
@@ -249,3 +256,22 @@ async def async_migrate_entry(hass: HomeAssistant, config_entry: ConfigEntry) ->
     _LOGGER.info("Migration to version %s successful", config_entry.version)
 
     return True
+
+
+async def async_remove_device_old(hass: HomeAssistant, entry: ConfigEntry):
+    """Remove obsolete source entity devices."""
+    entity_registry = er.async_get(hass)
+    device_registry = dr.async_get(hass)
+
+    source_entity = entity_registry.async_get(entry.options[CONF_SOURCE_SENSOR])
+    source_device_id = None
```

Comment:
If source_device_id is None, you don't have to do the loop and can just return. 

And so that's also why you don't need to iterate over the devices linked to the config entry. You can just get the device from the device registry as you know the ID.

Reply:
You can even just call this directly as you know the ID?

```
            device_registry.async_update_device(
                device_entry.id, remove_config_entry_id=entry.entry_id
            )
```